lib/codegen/src/simple_preopt.rs

```diff
@@ -536,11 +542,279 @@ fn simplify(pos: &mut FuncCursor, inst: Inst) {
     }
 }
 
+struct BranchOptInfo {
+    br_inst: Inst,
+    cmp_arg: Value,
+    destination: Ebb,
+    args: ValueList,
+    kind: BranchOptKind,
+}
+
+enum BranchOptKind {
+    EqualZero,
+    NotEqualZero,
+}
+
+fn branch_opt(pos: &mut FuncCursor, inst: Inst) {
+    let info = match pos.func.dfg[inst] {
+        InstructionData::BranchInt {
+            opcode: Opcode::Brif,
+            cond: br_cond,
+            destination,
+            ref args,
+        } => {
+            let first_arg = {
+                let args = pos.func.dfg.inst_args(inst);
+                args[0]
+            };
+            if let ValueDef::Result(iconst_inst, _) = pos.func.dfg.value_def(first_arg) {
+                if let InstructionData::BinaryImm {
+                    opcode: Opcode::IfcmpImm,
+                    imm: cmp_imm,
+                    arg: cmp_arg,
+                } = pos.func.dfg[iconst_inst]
+                {
+                    let cmp_imm: i64 = cmp_imm.into();
+                    if cmp_imm != 0 {
+                        return;
+                    }
+
+                    match br_cond {
+                        IntCC::NotEqual => BranchOptInfo {
+                            br_inst: inst,
+                            cmp_arg: cmp_arg,
+                            destination: destination,
+                            args: args.clone(),
+                            kind: BranchOptKind::NotEqualZero,
+                        },
+                        IntCC::Equal => BranchOptInfo {
+                            br_inst: inst,
+                            cmp_arg: cmp_arg,
+                            destination: destination,
+                            args: args.clone(),
+                            kind: BranchOptKind::EqualZero,
+                        },
```

Comment:
The duplication between the NotEqual and Equal arms here can be factored out by doing `let kind = match br_cond { ... };` to compute just the kind, and then you can do a single `BranchOptInfo { ... }`.